gpu/containers/include/pcl/gpu/containers/device_array.h

```diff
@@ -166,15 +166,15 @@ namespace pcl
               * \param rows number of rows to allocate
               * \param cols number of elements in each row
               * */
-            DeviceArray2D(int rows, int cols);
+            DeviceArray2D(std::size_t rows, std::size_t cols);
```

Comment:
Supply a deprecated replacement for class functions too (wherever the function signature changes. For return value change, nothing extra needs to be done)

Reply:
Is it necessary here? As with negative values it was a invalid usage before so we just increase the amount of cols and rows which can be used. As I don't believe someone use `-Wsign-compare` together with PCL, we should not break anything.

Reply:
Addition: How you want deprecate this btw? I adjust the member types, so to keep both variants I had to duplicate the members and synchronize them - I don't believe we want this.

Reply:
Deprecation allows API compatibility, even though the ABI is broken. The breakage is usually observed when someone accesses the function via a function pointer or `std::function`. Even though every minor release of PCL is breaking, going through deprecation gives time and warning to the user.

How to deprecate:
1. Add new functions (done)
2. Add deprecated to existing functions
3. Route old functions via the new functions in a safe manner if possible (here: yes)

Reply:
Any idea how to handle it for `ptr` ?

If I have:

```
template<class T>
PCL_DEPRECATED(...)
const T* ptr(int y_arg = 0) const;

template<class T>
const T* ptr(size_t y_arg = 0) const;
```

The compiler does not know what to use here, due to the default parameter:
https://github.com/PointCloudLibrary/pcl/blob/cd3ecd057afe6902296696b94af59b3f608f6a20/gpu/containers/include/pcl/gpu/containers/impl/device_memory.hpp#L94

Reply:
Best option here is to remove the default from the int overload.

Reply:
Ok, wasn't sure if this breaks the API, too.

Nevertheless: Currently I'm getting:
```
/home/sunblack/dev/pcl/gpu/surface/src/convex_hull.cpp:46:14: error: call to member function 'create' is ambiguous
  verts_inds.create(3, buffer_size);  
  ~~~~~~~~~~~^~~~~~
/home/sunblack/dev/pcl/gpu/containers/include/pcl/gpu/containers/device_array.h:206:18: note: candidate function
            void create(int rows, int cols);
                 ^
/home/sunblack/dev/pcl/gpu/containers/include/pcl/gpu/containers/device_array.h:212:18: note: candidate function
            void create(std::size_t rows, std::size_t cols);
```

Reasons: `buffer_size` is `std::size_t`, but `3` seems to be an `int`. So we cause the user via this both overloading to use `3ULL` instead of just `3` - do we really want this?

Reply:
That's unfortunate side effect of `C int promotion rules`, and `static_cast` is the way to go. The alternative is to break the API without deprecation.

Reply:
static_cast for variables would be okay, but as numeric values in code will be interpreted as int values by default I believe we will break with the deprecation more users code as without, as calling `create` with numbers (as e.g. `create(3, buffer_size)`) will be more common than using this function via a function pointer. So imho in this case a breaking API change is the better way.

Reply:
Added `feedback` to get more opinions. This seems like a rock and hard-place situation. Either break API or break interface.

Reply:
@SergioRAgostinho or  @taketwo Any opinion?

Reply:
I'm taking a step to ask if we should conform to `index_t` instead. That's why we added it right?

That being said what about the following [approach](https://godbolt.org/z/qDHUQc).

```cpp
std::size_t function(std::size_t n1, std::size_t n2) {
    return n1 * n2;
}

template<typename T>
[[deprecated]]
T function(T n1, T n2) {
    return function(static_cast<std::size_t> (n1), static_cast<std::size_t> (n2));
}


int main ()
{
    int i = 3;
    int ci = 3;
    std::size_t s = 3;
    const std::size_t cs = 3;

    // no deprecation warnings
    function(3, s);
    function(3, cs);

    // deprecation warnings
    function(3, 3);
    function(3, i);
    function(3, ci);
}
```

Reply:
Didn't it break the ABI, when you switch to a template function?

Reply:
I've read before that adding removing virtual functions breaks the ABI because the size of the vtable changes. In this case, I'm just adding normal methods so I'm tempted to say no. But then again, there's no point in worrying about that because the ABI will be broken regardless. The internal members of the class changed size.

Reply:
👍 for both ideas
* `index_t` for parameter type
* templated function + function for interface